Properly escape any HTML in string literals in Scaladoc examples

OberstK opened the pull request:
* SafeTagMarker '\u000E' is introduced on parsing of code blocks within Wiki comments
* CommentFactoryBase.scala was changed to not hand over this marker to the renderer of the ScalaDoc Document

Fixes https://github.com/scala/bug/issues/11424

Topics open to be discussed before merge:

* Do I overlook a dependency here? The Marker is deliberately introduced to enable jumping within a code block and clean out unwanted tags. I simply remove it after this is done.

I added a respective test for it as well to the partest directories

diesalbla commented:
I have noticed that half the changes in this PR (over 50 lines) are whitespace changes, not actual functionality changes. This may be a good point to ask: what is the policy regarding such changes? I have found that those are often introduced not by the programmer but by the editor or IDE, of which there are a lot to chose from. Once added, they add noise to the version control history. 

In my humble opinion, It would be better to remove those changes from this PR.

diesalbla reviewed (changes requested):
Opening the discussion on purely formatting changes.

OberstK commented:
Thanks for the feedback! The whitespace changes are definitely not intended and I will remove them as they were introduced by the auto formatter of my ide. I should have checked that better.

The parenthesis changes were intentionally and out of good spirit to bring the file at hand more in line with what I saw elsewhere in the code of scala (Boy Scout rule) but I can understand that this can also be done in a seperate PR.

Overall I am also quite interested if we want to introduce a global policy or even a linter to the project. This would like polite the history once with big format changes but could lead to an easier way of work for everyone.

I will remove the pure whitespace changes for sure and then wait for the discussion to be had how we handle the rest. Ok for you?

OberstK commented:
I will also check why the build fails on my partest. It seemed fine locally.

OberstK committed:
Clean up file at hand to prepare changes
* added several simplifications of the syntax
* changed mutable Map handling to reflect 2.13 deprecations

OberstK committed:
Remove artificial safeTagMarker in rendered StringLiterals

* SafeTagMarker '\u000E' is introduced on parsing of code blocks within Wiki comments
* CommentFactoryBase.scala was changed to not hand over this marker to the renderer of the ScalaDoc Document
* Added a test to check for leftover safeTagMarkers in Code Blocks in Comments

OberstK commented:
I updated the PR to remove not needed whitespace and indent changes and add the fix for the added Tests

I can still trim done the separate cleanup commit to not change parenthesis in the file
I think the changes to Map concat and two other deprecated methods is a usefull improvement for 2.13

NthPortal commented on src/scaladoc/scala/tools/nsc/doc/base/CommentFactoryBase.scala:
```diff
@@ -254,16 +254,16 @@ trait CommentFactoryBase { this: MemberLookupBase =>
           case Some(key) =>
             val value =
-              ((tags get key): @unchecked) match {
+              (tags get key: @unchecked) match {
```
can someone who knows the spec better confirm that `@unchecked` here applies to `tags get key` and not just `key`?

lrytz replied:
```
scala> import annotation.unchecked
import annotation.unchecked

scala> 1 + 2: @unchecked // print
   ((1).+(2): @unchecked) // : Int(3) @unchecked
```

Add `// print` in the repl and hit <kbd>tab</kbd> to show a pretty-print of the syntax tree

OberstK replied:
I see. I will check again what impact that has. From what I saw in the code, the unchecked annotation continued to make sense here without the parentheses.

OberstK replied:
I checked the specific code part and from my side the critical compiler warning in the pattern matching on tags get key is correctly suppressed. From what I see, the `@unchecked` flag is correctly applied without the parenthesis.

Atry replied:
I think `tags.get(key): @unchecked` is a more common sytle nowaday.

NthPortal commented on src/scaladoc/scala/tools/nsc/doc/base/CommentFactoryBase.scala:
```diff
@@ -254,16 +254,16 @@ trait CommentFactoryBase { this: MemberLookupBase =>
                 case None => oops("lastTagKey set when no tag exists for key")
               }
-            parse0(docBody, tags + (key -> value), lastTagKey, ls, inCodeBlock = true)
+            parse0(docBody, tags ++ Map(key -> value), lastTagKey, ls, inCodeBlock = true)
```
I don't think this needs to be wrapped in a `Map` (there are a few of these)

OberstK replied:
I added this to react to “+” being a deprecated function in 2.13

I tried using other ways to add this key value pair to the mutable map, but it seemed to me that this was prevented due to the @unchecked nature of tags (not sure)

I will give that another try and see if I can find a more elegant solution to replace “.+”

Maybe it would actually be beneficial to see if we can go for immutable Map here completely?

OberstK replied:
I changed these code parts to instead use an immutable map for the whole pattern matching clause and the usage of tags in this function to react to the deprecation of "+" on the generic Map trait. See my latest commit on the PR branch for this

Is that a valid solution for you?

NthPortal replied:
Yup, sounds good. `+` always creates a new `Map` regardless of whether or not it's mutable, so it may as well be immutable (which is at least optimised for such updates).

OberstK committed:
Changed the handling of Map additions to incorporate the 2.13 deprecation of + on the Map trait and instead used an immutable Map

Checked the usage of @unchecked on tags get key and verified that it works as intended to prevent compiler warnings on the non-check of Some(Nil) in the pattern match on tags get key

Changes to a StringBuild and a function to prevent deprecation warnings on the usage of Number/String concats and methods/functions without a clear return type

OberstK commented:
@diesalbla I force pushed the branch for this PR because of the formatting changes and the overall bad quality of the first commits. Please have look when you have the time to tell me if the new variant of this PR fits your expectation

OberstK commented:
Is there still something I can do about this PR that is missing or did just fell under the radar as it’s just scaladoc related :)

NthPortal commented:
@OberstK I suspect that it's just that it doesn't urgently need to get into the 2.13.0 release, so it's on the backburner for now.

SethTisue commented:
sorry about the long delay! glad to have this for 2.13.1
